core/src/main/scala/nl/vroste/zio/kinesis/client/Producer.scala

```diff
@@ -104,14 +110,53 @@ final case class ProducerSettings(
   failedDelay: Duration = 100.millis,
   metricsInterval: Duration = 30.seconds,
   updateShardInterval: Duration = 30.seconds,
-  aggregate: Boolean = false,
   allowedErrorRate: Double = 0.05,
-  shardPredictionParallelism: Int = 8
+  shardPrediction: Producer.ShardPrediction = Producer.ShardPrediction.Enabled(8),
+  aggregation: Producer.Aggregation = Producer.Aggregation.Disabled,
+  batchDuration: Option[Duration] = None
 ) {
   require(allowedErrorRate > 0 && allowedErrorRate <= 1.0, "allowedErrorRate must be between 0 and 1 (inclusive)")
+
+  aggregation match {
+    case Producer.Aggregation.ByPredictedShard(_) =>
+      require(aggregation != Producer.Aggregation.Disabled, "Aggregation requires shard prediction to be enabled")
+    case Producer.Aggregation.Disabled            => ()
+  }
+
+  shardPrediction match {
+    case Producer.ShardPrediction.Enabled(parallelism) =>
+      require(parallelism > 0, "shardPredictionParallelism must be > 0")
+    case Producer.ShardPrediction.Disabled             => ()
+  }
 }
 
 object Producer {
+  sealed trait ShardPrediction extends Product with Serializable
+  object ShardPrediction {
+    case object Disabled                           extends ShardPrediction
+    final case class Enabled(parallelism: Int = 8) extends ShardPrediction
+  }
+
+  sealed trait RichShardPrediction extends Product with Serializable
+
+  object RichShardPrediction {
+    case object Disabled extends RichShardPrediction
+    final case class Enabled(
+      parallelism: Int,
+      throttler: ShardThrottler,
+      md5Pool: ZPool[Nothing, MessageDigest],
+      shards: Ref[ShardMap],
+      triggerUpdateShards: UIO[Unit]
+    ) extends RichShardPrediction
+  }
+
+  sealed trait Aggregation extends Product with Serializable
+
+  object Aggregation {
+    case object Disabled                                                            extends Aggregation
+    final case class ByPredictedShard(aggregationDuration: Option[Duration] = None) extends Aggregation
```

Comment:
I could imagine it being useful to allow aggregation by partition key. This would have less aggregation than on the shard level, but doesn't have any issues with things potentially getting send to the wrong shard if the shardmap is outdated.

Out of scope for this pr though.